core/services/ocr2/plugins/ocr2keeper/evm21/registry.go

```diff
@@ -926,3 +937,75 @@ func (r *EvmRegistry) getBlockHash(blockNumber *big.Int) (common.Hash, error) {
 
 	return block.Hash(), nil
 }
+
+func (r *EvmRegistry) getTxBlock(txHash common.Hash) (*big.Int, common.Hash, error) {
+	txr, err := r.client.TransactionReceipt(r.ctx, txHash)
+	if err != nil {
+		return nil, common.Hash{}, err
+	}
+
+	return txr.BlockNumber, txr.BlockHash, nil
+}
+
+func (r *EvmRegistry) getCheckResult(p ocr2keepers.UpkeepPayload, failureReason uint8) ocr2keepers.CheckResult {
+	return ocr2keepers.CheckResult{
+		FailureReason: failureReason,
+		UpkeepID:      p.UpkeepID,
+		Trigger:       p.Trigger,
+		WorkID:        p.WorkID,
+		FastGasWei:    big.NewInt(0),
+		LinkNative:    big.NewInt(0),
+	}
+}
+
+// verifyCheckBlock checks that the check block and hash are valid
+func (r *EvmRegistry) verifyCheckBlock(checkBlock, upkeepId *big.Int, checkHash common.Hash, p ocr2keepers.UpkeepPayload, i int, results []ocr2keepers.CheckResult) bool {
+	// verify check block number is not too old
+	if r.bs.latestBlock-checkBlock.Int64() > validCheckBlockRange {
+		r.lggr.Warnf("latest block is %d, check block number %s is too old for upkeepId %s", r.bs.latestBlock, checkBlock, upkeepId)
+		results[i] = r.getCheckResult(p, UPKEEP_FAILURE_REASON_CHECK_BLOCK_TOO_OLD)
+		return false
+	}
+	// verify check block number and hash are valid
+	h, ok := r.bs.blocks[checkBlock.Int64()]
```

Comment:
accessing the map of block subscriber could lead to a data race

Reply:
done